### ISSUE TITLE
Rails integration?

### ISSUE THREAD

**Issue by jacobbednarz** [github/gh-ost#382]
We're coming from LHM and one of the good things about it is it's Rails integration which enables developers to write their migrations and have it run later on via our deployment tooling.

Is this something that you would see on the (GitHub or) gh-ost roadmap? I'm still in two minds about it since LHM is pretty straight forward whereas gh-ost gives you a bunch of power which you'd need to abstract into some form of chatops or UI. Not impossible, just unsure if it would be worth it. 

Right now most our migrations are being manually kicked off using gh-ost following a bit of a runbook since it's new and we're still treading very carefully. 

I'd be interested to hear your thoughts and workflows if you're able to share it.

**Comment by shlomi-noach** [comment]
I'd like to ping my colleague @arthurnn on this, since he's the maintainer of LHM. I'm unsure when he will see this as he's OOO at this time. @arthurnn I'm happy to hear your thoughts on this.

Generally speaking it would feel somewhat strange for me to have a rails wrapping around `gh-ost`, which is `golang` code. Why rails? And if rails, why not every other possible programming language? An answer to that may well be "because LHM exists"; and the fact GitHub uses rails also stands in favor.

---

@jacobbednarz we are also kicking migrations manually at this time, but on the verge of having a fully automated (with human safety latches) cycle. This begins with inception of the schema change, using `structure.sql` to generate the migration command, through human approvals, through queuing into execution, execution itself, human `cut-over`, then final cleanup and updates. It ties very tightly into GitHub's internal flow of using PRs for schema changes.

I'm writing this very briefly because this is still ongoing and we're surely to elaborate once this is out (though this will not be an open source solution as it is so strongly coupled with our internals).

**Comment by jacobbednarz** [comment]
> Generally speaking it would feel somewhat strange for me to have a rails wrapping around `gh-ost`, which is golang code.

I would see it less wrapping the code and more as a nice abstraction to kick off a migration. I.e. something like the following in rails migration.

```rb
class MyExampleMigration < ActiveRecord::Migration
  def up
    Ghost.drop_table :my_old_table
  end
end
```

This would then generate the `gh-ost` command to run through whatever process you would usually do.

> Why rails? And if rails, why not every other possible programming language? An answer to that may well be "because LHM exists"; and the fact GitHub uses rails also stands in favor.

After thinking about this more, I realise this is more of a "it's always been this way (with LHM) so let's keep doing this" - which could be completely wrong and it might be time to take a new approach to running migrations.

For the record, our flow looks like this:

- PR the changes (migration + code are separate if there is a DB change dependency)
- Migration is deployed
- Migration is run using LHM
- PR code change is deployed

Perhaps it's time we look to handle migrations completely separately using `gh-ost`? On a side note, are your `gh-ost` migrations version controlled? Or just pulled off the `structure.sql`?

**Comment by shlomi-noach** [comment]
> On a side note, are your gh-ost migrations version controlled? 

Can you elaborate on that? I'm not sure I follow.

A `gh-ost` run is coupled with a PR. That PR can have multiple changes on multiple tables. A `gh-ost` execution can fail (e.g. `critical-load` is met) and then we kick it again, manually at this time, and automatically in the future. There is no versioning to a single `gh-ost` execution.

The result of a `gh-ost` run is an updated schema in production, which then makes (given `git` some magic) CI tests go green, making the PR mergable.

**Comment by jacobbednarz** [comment]
Sure, let me rephrase it. 

At the moment, we commit our schema changes using `structure.sql` and the ActiveRecord migration file (similar the [example above](https://github.com/github/gh-ost/issues/382#issuecomment-286659315)). However my understanding with `gh-ost` is that since it's outside of Rails, the ActiveRecord migration is of little use (at least in production). Once in production, a rake task is kicked off that pretty much boils down to a custom version of `rake db:migrate`.

> The result of a gh-ost run is an updated schema in production, which then makes (given git some magic) CI tests go green, making the PR mergable.

Ah. This is quite different to our flow. So you put it all in one PR, then through some CI process kick off the migration behind the scenes and then the code itself cannot be merged until it's been successfully run. What does the file you commit look like for the `gh-ost` migration?

**Comment by kirs** [comment]
I came across this issue and I wanted to mention that I'd also be interested in learning about the best practice of integrating gh-ost and Rails migrations.

I think the better question to ask is: how does Github embed gh-ost into the Rails monolith? Is there a supervisor that kicks off and monitors the gh-ost process? Does this supervisor run outside of Rails deployment?

**Comment by shlomi-noach** [comment]
Some answers will be covered by https://www.percona.com/live/17/sessions/automating-schema-changes-using-gh-ost, and this is an ongoing work.

TL;DR, yes, an external supervisor to kick off `gh-ost` independently of the `rails` deployment.

**Comment by jacobbednarz** [comment]
I've started doing a bit of tooling to do with rails integration (we're going from LHM to `gh-ost`) and the TL;DR of our flow looks like this:

- Open a feature branch with the usual DB ActiveRecord migration included.
- Generate the `gh-ost` command using a CI step and post it via to the pull request for verification. Happy to share the script we use for this if you'd like to eyeball it.
- Run a series of manual `--test-on-replica` and `--migrate-on-replica` runs and perform some data integrity checks using `pt-table-checksum`, a hash of the query results and anything else you might need to check based on the migration (for example, indexes).
- Once happy, migrate the master using `gh-ost`.
- Manually insert migration values into `schema_migrations` table.
- Merge the pull request with migration changes.
- Automatically clean up `*_del` after a defined period of time.

We haven't automated the runs yet because we are still getting a feel for how our infrastructure and edge cases (mainly RDS) are handled but eventually, we will be hooking a bunch of this up to our CI steps similar to what I think GitHub are doing in their pipeline.

@kirs If you're interested in more details on how we're doing it, I'm happy for you to shoot me an email (Simon Eskildsen should be able to rustle it up if you) and I can hand over whatever you're after. I try to put some of our findings back here but it's hard to work out at the moment what is benefitting others and what would just be noise.

**Comment by shlomi-noach** [comment]
cc @tomkrouper who is similarly working on this on our side.

**Comment by kirs** [comment]
@jacobbednarz thanks for sharing the plan! I have been looking a bit into how to fit gh-ost into Rails migrations but I don't think we'll be moving forward any time soon.

**Comment by tomkrouper** [comment]
> Generate the gh-ost command using a CI step and post it via to the pull request for verification. Happy to share the script we use for this if you'd like to eyeball it.

We have a separate command for this part, but all the rest of the steps sound very similar to what we do.

**Comment by jacobbednarz** [comment]
@kirs Are there any blockers or issues with gh-ost itself for you not proceeding? If there are, it would be great to iron them out (or at least get them into issues) since this tooling has given us far more confidence and flexibility in our DB migrations and it would be great to have more people in that position.

**Comment by kirs** [comment]
There are no blockers with gh-ost that I know. It's just that teams are currently busy with some other things and we didn't face into critical troubles with LHM at this point.

**Comment by jacobbednarz** [comment]
That's good to hear @kirs 😄  Hopefully when/if you get back to this, I have a more stable process to share.

**Comment by shuhaowu** [comment]
@jacobbednarz can you share your process? I've given this some thought already and come up with a rudimentary design as well that would work but don't right now have the time to work on it. Would be interesting to see how it differs from yours.

**Comment by jacobbednarz** [comment]
@shuhaowu Our process is pretty much https://github.com/github/gh-ost/issues/382#issuecomment-295023443 but I'm happy to share more details on a particular step if you'd like?

**Comment by gaurish** [comment]
@jacobbednarz 
can you share this step:

> Generate the gh-ost command using a CI step and post it via to the pull request for verification. Happy to share the script we use for this if you'd like to eyeball it.

**Comment by jacobbednarz** [comment]
Sure @gaurish. Within CI we have a step that checks the PR for migrations and if it is found runs the [`generate_ghost_command`](https://gist.github.com/jacobbednarz/57fc7313c0ea3d89d3f4c36b655210e6) against that file. 

Here is an example migration and it's output.

```rb
# db/migrate/1234_remove_username_from_users.rb
class RemoveUsernameFromUsers < ActiveRecord::Migration
  def up
    remove_column :users, :username
  end

  def down
    add_column :users, :username, "varchar(255) DEFAULT NOT NULL"
  end
end
```

Output

```sh
gh-ost \
--user="ghost" \
--password="PASSWORD" \
--port=3306 \
--host="DB_HOST" \
--database="DB_NAME" \
--table="users" \
--alter="DROP \`username\`" \
--assume-rbr \
--hooks-path="hooks" \
--panic-flag-file=/tmp/ghost.panic.flag \
--postpone-cut-over-flag-file=/tmp/ghost.postpone.flag \
[--allow-on-master | --migrate-on-replica] \
# --execute
```

That output is then posted back to the pull request using the GitHub API. You can then trigger a `gh-ost` run on our DB migration cluster which allows you to test the migration and run some sanity and integrity checks against it using [`gh-ost` hooks](https://github.com/github/gh-ost/blob/master/doc/hooks.md).

For production, we then use a [Buildkite block step](https://buildkite.com/docs/pipelines/block-step) to confirm the migration output is as expected and then it is run against our production DB.

Note: The output above assumes _our_ sane defaults like running on AWS RDS. I'd recommend consulting the documentation before running it blindly. The `generate_ghost_command` script is also pretty simple but works for our needs so YMMV.

**Comment by mahemoff** [comment]
I've made an initial attempt at a library for Rails integration. Worked for my own needs so far, at least ... feedback welcome.

https://github.com/playerfm/GhostMigration
